src/ChunkDef.h

```diff
@@ -40,6 +40,9 @@ using cEntityList = std::vector<OwnedEntity>;
 /** The datatype used by blockdata */
 typedef unsigned char BLOCKTYPE;
 
+/** The datatype used by block metadata */
+typedef unsigned char BLOCKMETATYPE;
+
 /** The datatype used by nibbledata (meta, light, skylight) */
```

Comment:
To what?

Reply:
ChunkDataSerializer sends blocks. They've always had 8 bit metadata of which only 4 bits are used. Nothing has changed there. (Although there may be a case to be argued that it should mask the meta to avoid possible future surprises.)

Banners aren't blocks though. They're block entities. Block entities are responsible for their own updates via SendUpdateBlockEntity and SendBlockChange.

SendUpdateBlockEntity isn't affected. There are no material changes and banner colour is always sent as an int value in the NBT blob.

SendBlockChange is the one that uses meta. >= 1.13 uses the given meta as-is (colour and rotation where the colour part gets promoted into the mapped block id), < 1.13 masks to get only the lowest 4 bits (rotation).

So all is fine. (Or at least: it works for me :-) )

Reply:
Or, to put it another way, block meta and block _entity_ meta aren't quite the same. The chunk holds the block type (standing banner or wall banner) and a meta that represents the rotation. Block entities take a copy of that and pass their version through the update entity protocol path. In the case of banners the meta is modified to include colour so that the block id mapping can do the right thing. The meta in the chunk is fixed when the block is placed and entities never write their copy back to the chunk.